### ISSUE TITLE
False positive for restrict-plus-operands rule - Readonly and generic type

### ISSUE THREAD

**Issue by madprog** [palantir/tslint#3902]
### Bug Report

- __TSLint version__: 5.9.0, 5.1.0, 13723b506b8d0531312b07188fa75cb296a055d9
- __TypeScript version__: 2.8.3
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
interface Struct {
    value: number;
}

function factory<S extends Struct>(struct: S, callback: (prevState: Readonly<S>) => void): void { }

class MyClass {
    readonly value = 0;
    readonly foo = factory(this, (struct) => {
        const a = struct.value + 1;
    });
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "restrict-plus-operands": true
  }
}
```

#### Actual behavior
`Operands of '+' operation must either be both strings or both numbers`

#### Expected behavior
Should not report an error since `struct.value` is a number too.

#### Additional notes
The error disappears if `Readonly<S>` is replaced with `S` in the type of the `callback` argument.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
